### PR TITLE
 Scheduling support for warehouse 

### PR DESCRIPTION
**Pull request by psrikanth88**
This PR also includes few changes like adding timings to wh_uploads to track start and end time at more granular level and also store count of rows uploaded in each table.

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-server/234)

### CODE REVIEW

File: warehouse/warehouse.go [hunk 1]
```diff
@@ -447,13 +470,135 @@ func isDestInProgress(warehouse warehouseutils.WarehouseT) bool {
...
+		pos = idx - 1
+		break
+	}
+
+	// if current time is less than first start time in a day, take last start time in prev day
+	if pos < 0 {
+		return startOfDay(now).Add(time.Hour * time.Duration(-24)).Add(time.Minute * time.Duration(allStartTimes[len(allStartTimes)-1]))
+	}
+	return startOfDay(now).Add(time.Minute * time.Duration(allStartTimes[pos]))
+}
```
Comment (fxenik) [inline warehouse/warehouse.go:572]: I think there is a bug here: 
if current time is after last value of allStartTimes (e.g last value is 22:00 and current time is 23:00), pos variable will never be set based on idx, and will remain 0. This will result in adding first value's minutes to current start of day.

I suggest you create a small test for this function specifically, as it is easy to miss edge cases when times are close to day's starts/ends (I now I did have similar bugs in similar Blendo's code :)

Keep in mind, testing this function will require a small refactoring passing the current time as argument, instead of fixing it to time.Now()

File: warehouse/warehouse.go [hunk 2]
```diff
@@ -447,13 +470,135 @@ func isDestInProgress(warehouse warehouseutils.WarehouseT) bool {
...
+
+func (wh *HandleT) canStartUpload(warehouse warehouseutils.WarehouseT) bool {
+	syncFrequency := warehouseutils.GetConfigValue(warehouseutils.SyncFrequency, warehouse)
+	syncStartAt := warehouseutils.GetConfigValue(warehouseutils.SyncStartAt, warehouse)
+	if syncFrequency != "" && syncStartAt != "" {
+		prevScheduledTime := getPrevScheduledTime(syncFrequency, syncStartAt)
+		lastUploadExecTime := wh.getLastUploadStartTime(warehouse)
+		// start upload only if no upload has started in current window
+		// eg. with prec scheduled time 14:00 and current time 15:00, start only if prev upload hasn't started after 14:00
```
Comment (fxenik) [inline warehouse/warehouse.go:595]: typo in prec

File: warehouse/warehouse.go [hunk 3]
```diff
@@ -447,13 +470,135 @@ func isDestInProgress(warehouse warehouseutils.WarehouseT) bool {
...
+// prev scheduled time for current time (eg. 18:00 -> 16:00 same day, 00:30 -> 22:00 prev day)
+func getPrevScheduledTime(syncFrequency, syncStartAt string) time.Time {
+	allStartTimes := scheduledTimes(syncFrequency, syncStartAt)
+
+	loc, _ := time.LoadLocation("UTC")
+	now := time.Now().In(loc)
+	// current time in minutes since start of day
+	currMins := now.Hour()*60 + now.Minute()
+
+	// get position where current time can fit in the sorted list of allallStartTimes
```
Comment (fxenik) [inline warehouse/warehouse.go:557]: typo in allallStartTimes

File: warehouse/utils/utils.go [hunk 4]
```diff
@@ -187,10 +193,84 @@ func GetSchemaDiff(currentSchema, uploadSchema map[string]map[string]string) (di
...
+		if idx < len(fields)-1 {
+			columns += ","
+		}
+		values = append(values, f.Value)
+	}
+	sqlStatement := fmt.Sprintf(`UPDATE %s SET %s WHERE id=$1`, warehouseUploadsTable, columns)
+	_, err = dbHandle.Exec(sqlStatement, values...)
 	if err != nil {
 		panic(err)
 	}
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/utils/utils.go:276]: removed it and used SetUploadColumns everywhere

File: warehouse/utils/utils.go [hunk 5]
```diff
@@ -187,10 +193,84 @@ func GetSchemaDiff(currentSchema, uploadSchema map[string]map[string]string) (di
...
+		if !ok {
+			continue
+		}
+		timing := map[string]string{}
+		for status, t := range m {
+			timing[status] = t.(string)
+		}
+		timings = append(timings, timing)
+	}
+	return
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/utils/utils.go:227]: actually removed this part as part of the SetUploadColumns change, since it can be directly scanned to a []map[string]string structure

File: warehouse/warehouse.go (hunk 2 above)
Comment (psrikanth88) [reply warehouse/warehouse.go:595]: changed

File: warehouse/warehouse.go (hunk 3 above)
Comment (psrikanth88) [reply warehouse/warehouse.go:557]: changed

File: warehouse/warehouse.go (hunk 1 above)
Comment (psrikanth88) [reply warehouse/warehouse.go:572]: fixed that and added unit tests for the func

File: warehouse/warehouse.go [hunk 6]
```diff
@@ -447,13 +470,135 @@ func isDestInProgress(warehouse warehouseutils.WarehouseT) bool {
...
+		if mins < 0 {
+			break
+		}
+		prependTimes = append(prependTimes, mins)
+		counter++
+	}
+	times = append(funk.ReverseInt(prependTimes), times...)
+	scheduledTimesCache[fmt.Sprintf(`%s-%s`, syncFrequency, syncStartAt)] = times
+	return
+}
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/warehouse.go:540]: moved it

File: warehouse/utils/utils.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/utils/utils.go:227]: although need to come up with a json utils lib to handle all the json req's in the code. Haven't come up with anything clean, so didn't do it for this change

File: warehouse/utils/utils.go [hunk 7]
```diff
@@ -240,34 +231,32 @@ func GetNewTimings(upload UploadT, dbHandle *sql.DB, status string) []byte {
...
-	sqlStatement := fmt.Sprintf(`UPDATE %s SET status=$1, %s updated_at=$2, timings=$3 WHERE id=$4`, warehouseUploadsTable, extraFields)
-	_, err = dbHandle.Exec(sqlStatement, status, time.Now(), marshalledTimings, upload.ID)
-	if err != nil {
-		panic(err)
-	}
-	return
-}
-
 // SetUploadColumns sets any column values passed as args in UploadColumnT format for warehouseUploadsTable
 func SetUploadColumns(upload UploadT, dbHandle *sql.DB, fields ...UploadColumnT) (err error) {
```
Comment (fxenik) [inline warehouse/utils/utils.go:235]: My suggestion was not to merge SetUploadStatus and SetUploadColumns. SetUploadColumns should stay as it was, having just the job of updating the table. SetUploadStatus should be rewritten to call SetUploadColumns, after doing its specific tasks (i.e shorthand for status column and computing timings)

File: warehouse/warehouse.go [hunk 8]
```diff
@@ -367,8 +367,8 @@ func (wh *HandleT) initTableUploads(upload warehouseutils.UploadT, schema map[st
 }
 
 func (wh *HandleT) initUpload(warehouse warehouseutils.WarehouseT, jsonUploadsList []*StagingFileT) warehouseutils.UploadT {
-	sqlStatement := fmt.Sprintf(`INSERT INTO %s (source_id, namespace, destination_id, destination_type, start_staging_file_id, end_staging_file_id, start_load_file_id, end_load_file_id, status, schema, error, first_event_at, last_event_at, created_at, updated_at, timings)
```
Comment (fxenik) [inline warehouse/warehouse.go]: Why is timing initialisation removed ?

File: warehouse/warehouse_test.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
+package warehouse_test
```
Comment (fxenik) [inline warehouse/warehouse_test.go:1]: Consider renaming this file to scheduling_test.go ?

File: warehouse/utils/utils.go [hunk 10]
```diff
@@ -202,36 +209,20 @@ type UploadColumnT struct {
...
-	}
+	err = json.Unmarshal(rawJSON, &timings)
 	return
 }
 
 // GetNewTimings appends current status with current time to timings column
 // eg. status: exported_data, timings: [{exporting_data: 2020-04-21 15:16:19.687716] -> [{exporting_data: 2020-04-21 15:16:19.687716, exported_data: 2020-04-21 15:26:34.344356}]
 func GetNewTimings(upload UploadT, dbHandle *sql.DB, status string) []byte {
 	timings := GetUploadTimings(upload, dbHandle)
-	timing := map[string]string{status: time.Now().Format(misc.RFC3339Milli)}
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/utils/utils.go]: did this intentionally, so as to compare lastUploadTime with scheduled time. We are taking the scheduled time from the user in UTC. Did not want to complicate by asking the user for timezone (or decoding ourselves by his browser timezone)

File: warehouse/utils/utils.go (hunk 7 above)
Comment (psrikanth88) [reply warehouse/utils/utils.go:235]: got it, have both funcs and remvoed repetition of sql query

File: warehouse/warehouse.go (hunk 8 above)
Comment (psrikanth88) [reply warehouse/warehouse.go]: added initially to avoid error in sql Scan(). Since it is handled now, did not initialize it

File: warehouse/warehouse_test.go (hunk 9 above)
Comment (psrikanth88) [reply warehouse/warehouse_test.go:1]: renamed

File: warehouse/warehouse.go [hunk 11]
```diff
@@ -388,7 +388,7 @@ func (wh *HandleT) initUpload(warehouse warehouseutils.WarehouseT, jsonUploadsLi
 		lastEventAt = jsonUploadsList[0].LastEventAt
 	}
 
-	row := stmt.QueryRow(warehouse.Source.ID, namespace, warehouse.Destination.ID, wh.destType, startJSONID, endJSONID, 0, 0, warehouseutils.WaitingState, "{}", "{}", firstEventAt, lastEventAt, time.Now(), time.Now(), "[]")
+	row := stmt.QueryRow(warehouse.Source.ID, namespace, warehouse.Destination.ID, wh.destType, startJSONID, endJSONID, 0, 0, warehouseutils.WaitingState, "{}", "{}", firstEventAt, lastEventAt, timeutil.Now(), timeutil.Now())
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/warehouse.go:391]: :thum

File: warehouse/warehouse.go [hunk 12]
```diff
@@ -343,7 +343,7 @@ func (wh *HandleT) initTableUploads(upload warehouseutils.UploadT, schema map[st
 	}
 
 	for _, table := range tables {
-		_, err = stmt.Exec(upload.ID, table, "waiting", "{}", time.Now(), time.Now())
+		_, err = stmt.Exec(upload.ID, table, "waiting", "{}", timeutil.Now(), timeutil.Now())
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/warehouse.go:346]: 👍

File: warehouse/warehouse.go [hunk 13]
```diff
@@ -673,7 +561,7 @@ func (wh *HandleT) mainLoop() {
 				}
 				wh.uploadToWarehouseQ <- jobs
 			} else {
-				if !wh.canStartUpload(warehouse) {
+				if !wh.CanStartUpload(warehouse) {
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/warehouse.go:564]: missed it, changed it back
